libindy/src/services/crypto/mod.rs

```diff
@@ -43,22 +43,60 @@ impl CryptoService {
         }
     }
 
-    pub fn create_key(&self, key_info: &KeyInfo) -> Result<Key, CryptoError> {
-        let crypto_type_name = key_info.crypto_type
-            .as_ref()
-            .map(String::as_str)
-            .unwrap_or(DEFAULT_CRYPTO_TYPE);
+    pub fn get_crypto_type_name<'a>(crypto_type: &'a Option<String>) -> (&'a str) {
+        crypto_type.as_ref().map(String::as_str).unwrap_or(DEFAULT_CRYPTO_TYPE)
+    }
+
+    pub fn get_crypto_type_name_from_verkey<'a>(verkey: &'a String) -> (&'a str) {
+        if verkey.contains(":") {
+            let splits: Vec<&str> = verkey.split(":").collect();
+            splits[1]
+        } else {
+            DEFAULT_CRYPTO_TYPE
+        }
+    }
+
+    pub fn get_verkey_and_crypto_type_name_from_verkey<'a>(verkey: &'a str) -> (&'a str, &'a str) {
+        if verkey.contains(":") {
+            let splits: Vec<&str> = verkey.split(":").collect();
+            (splits[0], splits[1])
+        } else {
+            (verkey, DEFAULT_CRYPTO_TYPE)
+        }
+    }
+
+    pub fn is_crypto_type_supported(&self, crypto_type_name: &str) -> bool {
+        self.crypto_types.contains_key(crypto_type_name)
+    }
 
-        if !self.crypto_types.contains_key(crypto_type_name) {
+    fn get_crypto_type(&self, crypto_type_name: &str) -> &Box<CryptoType> {
+        self.crypto_types.get(crypto_type_name).unwrap()
+    }
+
+    pub fn get_crypto_name_and_keypair<'a>(&self, crypto_type: &'a Option<String>,
```

Comment:
> pub fn get_crypto_name_and_keypair<'a>(&self, crypto_type: &'a Option<String>,
>                                               seed: &'a Option<String>) -> Result<(&'a str, Vec<u8>, Vec<u8>), CryptoError>

This function mix a lot of responsibilities. So i suggest to remove this. but we can update ```get_crypto_type``` to cover the most of boilerplate without introducing single responsibility problems